src/runtime/thread_pool_common.h

```diff
@@ -0,0 +1,282 @@
+        halide_cond_init(&work_queue.wakeup_b_team);
+        work_queue.jobs = NULL;
+
+        // Compute the desired number of threads to use. Other code
+        // can also mess with this value, but only when the work queue
+        // is locked.
+        if (!work_queue.desired_num_threads) {
+            char *threads_str = getenv("HL_NUM_THREADS");
+            if (!threads_str) {
+                // Legacy name for HL_NUM_THREADS
+                threads_str = getenv("HL_NUMTHREADS");
+            }
+            if (threads_str) {
+                work_queue.desired_num_threads = atoi(threads_str);
+            } else {
+                work_queue.desired_num_threads = halide_host_cpu_count();
+            }
+        }
+        if (work_queue.desired_num_threads > MAX_THREADS) {
+            work_queue.desired_num_threads = MAX_THREADS;
+        } else if (work_queue.desired_num_threads < 1) {
+            work_queue.desired_num_threads = 1;
+        }
+        work_queue.threads_created = 0;
+
+        // Everyone starts on the a team.
+        work_queue.a_team_size = work_queue.desired_num_threads;
+
+        work_queue.initialized = true;
+    }
+
+    while (work_queue.threads_created < work_queue.desired_num_threads - 1) {
+        // We might need to make some new threads, if work_queue.desired_num_threads has
+        // increased.
+        work_queue.threads[work_queue.threads_created++] =
+            halide_spawn_thread(worker_thread, NULL);
+    }
+
+    // Make the job.
+    work job;
+    job.f = f;               // The job should call this function. It takes an index and a closure.
+    job.user_context = user_context;
+    job.next = min;          // Start at this index.
+    job.max  = min + size;   // Keep going until one less than this index.
+    job.closure = closure;   // Use this closure.
+    job.exit_status = 0;     // The job hasn't failed yet
+    job.active_workers = 0;  // Nobody is working on this yet
+
+    if (!work_queue.jobs && size < work_queue.desired_num_threads) {
+        // If there's no nested parallelism happening and there are
+        // fewer tasks to do than threads, then set the target A team
+        // size so that some threads will put themselves to sleep
+        // until a larger job arrives.
+        work_queue.target_a_team_size = size;
+    } else {
+        // Otherwise the target A team size is
+        // desired_num_threads. This may still be less than
+        // threads_created if desired_num_threads has been reduced by
+        // other code.
+        work_queue.target_a_team_size = work_queue.desired_num_threads;
+    }
+
+    // If there are fewer threads than we would like on the a team,
+    // wake up the b team too.
+    bool wake_b_team = work_queue.target_a_team_size > work_queue.a_team_size;
+
+    // Push the job onto the stack.
+    job.next_job = work_queue.jobs;
+    work_queue.jobs = &job;
+
+    halide_mutex_unlock(&work_queue.mutex);
```

Comment:
Pedantic style is to move it inside and I would do so unless benchmarking shows a performance hit. It should not affect correctness either way